### DISCUSSION TITLE
wayland 环境下 Qt 类程序的父子窗口定位问题

### DISCUSSION THREAD

**Discussion by yanhuacuo** [linuxdeepin/developer-center#5033] [Development Questions | 开发答疑] [answered]
看到 deepin V23 引入了 wayland 环境，我们这些 Qt 开发者也当早做打算，但是发现一个普遍性的问题，Qt 的消息弹窗，对话框弹窗，甚至所有子窗口弹出都是据根父级窗口的位置计算的。而 wayland 不支持窗口间获取这些数据。

比如，随便写个弹窗，让它吐个坐标，在 wayland 下永远是 （0,0）

````
    int x = this->geometry().x();
    int y = this->geometry().y();  
    qDebug() << "Window position: (" << x << ", " << y << ")";
````

唯今之计，就是在启动脚本中指定一个临时变量，强指 `xcb` 。

````
QT_QPA_PLATFORM=xcb ./testApp
````

这实际是调用 x11 的接口来运行了，终不是长久之计。而 Qt 认为这是 wayland 协议的缺陷（bug），并无修复计划。

看到 deepin 博客上写过，深度截图搞定了这问题。

所以，针对这个问题，Qt框架内，有无符规范性的解决办法? 

如果 Qt 放任这个问题，而 wayland 依然故我（他们一直这样），未来 deepin 怎么处理 Qt 程序这个兼容性问题?

盼复。

**Comment by Decodetalkers** [comment]
wayland有wayland的方法，你要拿到其他窗口都要通过窗口管理器去拿，窗口各自是独立的。

如果你不是做窗口管理器绑定的软件，就没必要去探索其他窗口的位置了

**Reply by yanhuacuo** [reply]
没写过 qt 软件?
现状是 : 一行单简的消息弹窗，弹窗位置都是飘移的啊。
https://github.com/cppxiaozhu/Demo

**Reply by Decodetalkers** [reply]
你这个消息就应该使用 org.freedesktop.Notifications 接口啊，不要自己去定位窗口

**Reply by Decodetalkers** [reply]
wayland下窗口位置放到哪里你要和wm进行沟通，尽量使用通用的api

**Reply by Decodetalkers** [reply]
如果你一定要自己去定位窗口的话，目前wayland可以定位8个位置，https://wayland.app/protocols/wlr-layer-shell-unstable-v1
这个协议，其他位置可以从左上开始设置位置的偏移调，但是这个协议在gnome上是不存在的，gnome的话我不了解

**Comment by yanhuacuo** [comment]
> 
https://github.com/cppxiaozhu/Demo/blob/82511d03d854f2fee45088c4b04e2e6b2abced8b/mainwindow.cpp#L20

这是一个常规的，通用的写法 : 按下 button，弹个消息框，没有用到底层协议。


类似的操作，在应用程序中很常见。但是这个写法在 wayland 下是失效的，会窗口飘移。

**Reply by Decodetalkers** [reply]
你如果想要一个dialog或者popup,应该将窗口属性设置成popup,你这个环境是gnome吧，gnome应该有gnome的处理，在kde或者sway还有deepin上对新窗口默认就是中间显示，我完全不知道gnome怎么处理的，你应该去咨询gnome的开发者。这些行为完全是wm自己定义的

**Reply by yanhuacuo** [reply]
与桌面环境无关的

**Reply by Decodetalkers** [reply]
kde和sway我刚才都试过了，位置都是中间，我认为是gnome做的事情

**Reply by Decodetalkers** [reply, marked as answer]
更新，deepin上弹窗位置是中间，也没有问题，看来你只能去问gnome了

看了下demo, QMessageBox 这个本来就是在父窗口中间位置弹窗的，这个没有问题，不需要定位全局的坐标，这个例子如果失败那就是gnome的问题

**Reply by yanhuacuo** [reply]
明白了，wayland 合成器的差异。

**Comment by BLumia** [comment]
wayland 设计如此，就是故意不让你获取窗口位置的。其实有一些合成器是有拓展协议去支持获取和设置窗口位置的，但并非所有 wayland 合成器下都可用。

QMessageBox 应该是对应到 wayland 的 popup 类型窗口，一般合成器会把这种窗口摆到它的父窗口的中间。

如果你希望你的逻辑可移植的话，对于实现通知是使用通知的接口，实现无边框窗口移动的话 Qt 也提供了一个 `startSystemMove/Resize()`，都可以在不知道窗口位置的前提下进行窗口的移动和调整大小。~~真的觉得窗口位置是必要的话，就该和 wayland 设计人员对喷了~~（x

（ps. 和你的问题无关：Qt 的 .user 文件是不需要 check in 到 VCS 里的）

**Reply by yanhuacuo** [reply]
测试了一下最新版kde，它的 wayland 合成器下，弹窗位置是正常的。4年前 qt论坛就有人反应过这个问题，qt员工很笃定地说这个是 wayland 的问题，他们qt这边，不准备做什么适应性改变。

**Reply by BLumia** [reply]
因为就是 wayland 问题。窗口位置的设置并不在 wayland 协议本身里。KDE 可能提供了拓展协议做这个事情（但我没查到对应的接口，只查到了像是给 plasma shell 用的接口）。总之由此原因，应用应当尽可能规避设置窗口位置的操作。即便 DDE 也扩展协议做这件事，那在非 DDE 与 KDE 的环境下也仍然会有可能无法获取位置的问题。

**Reply by yanhuacuo** [reply]
KDE 重写了 [sub-dialog](https://github.com/yanhuacuo/plasma-desktop/blob/f7283da3fafda707fe6668270d1dcb86862c2800/containments/desktop/plugins/folder/subdialog.cpp)

**Reply by yanhuacuo** [reply]
最大的影响就是弹窗消息都找不到他亲爹的位置，有点「过于离谱」。
